src/ai/backend/manager/types.py

```diff
@@ -82,75 +109,100 @@ class State(enum.Enum):
 
 @dataclass
 class TriState(Generic[TVal]):
+    """
+    TriState is a class that represents partial updates to an attribute of an object.
+    It is used to indicate whether an attribute should be updated, set to None, or not modified at all.
+    It can be in one of three states:
+    - UPDATE: The attribute should be updated with the given value.
+    - NULLIFY: The attribute should be set to None.
+    - NOP: No operation should be performed on the attribute.
+    """
+
     _attr_name: str
-    _state: State
+    _state: TriStateEnum
     _value: Optional[TVal]
 
-    def __init__(self, attr_name: str, state: State, value: Optional[TVal]):
+    def __init__(self, attr_name: str, state: TriStateEnum, value: Optional[TVal]):
         self._attr_name = attr_name
         self._state = state
         self._value = value
 
+    @classmethod
+    def from_graphql(cls, attr_name: str, value: Optional[TVal] | UndefinedType) -> TriState[TVal]:
+        if value is None:
+            return cls.nullify(attr_name)
+        if value is Undefined:
+            return cls.nop(attr_name)
+        return cls.update(attr_name, value)  # type: ignore
+
     @classmethod
     def update(cls, attr_name: str, value: TVal) -> TriState[TVal]:
-        return cls(attr_name, state=State.UPDATE, value=value)
+        return cls(attr_name, state=TriStateEnum.UPDATE, value=value)
 
     @classmethod
     def nullify(cls, attr_name: str) -> TriState[TVal]:
-        return cls(attr_name, state=State.NULLIFY, value=None)
+        return cls(attr_name, state=TriStateEnum.NULLIFY, value=None)
 
     @classmethod
     def nop(cls, attr_name: str) -> TriState[TVal]:
-        return cls(attr_name, state=State.NOP, value=None)
+        return cls(attr_name, state=TriStateEnum.NOP, value=None)
 
     def value(self) -> Optional[TVal]:
-        if self._state == State.UPDATE:
+        if self._state == TriStateEnum.UPDATE:
             return self._value
         raise ValueError(f"Value is not set for {self._attr_name}")
 
-    def state(self) -> State:
+    def state(self) -> TriStateEnum:
         return self._state
 
     def set_attr(self, obj: Any) -> None:
         match self._state:
-            case State.UPDATE:
+            case TriStateEnum.UPDATE:
                 setattr(obj, self._attr_name, self._value)
-            case State.NULLIFY:
+            case TriStateEnum.NULLIFY:
                 setattr(obj, self._attr_name, None)
-            case State.NOP:
+            case TriStateEnum.NOP:
                 pass
 
 
 class OptionalState(TriState[TVal]):
-    def __init__(self, attr_name: str, state: State, value: Optional[TVal]):
+    """
+    OptionalState is a class that represents partial updates to an attribute of an object.
```

Comment:
Just a question, are we going to use the name `OptionalState`?

Reply:
I chose the option with the least modifications for now.